### PR TITLE
[wip][LA-275] add leapfrog-multinode-aio into gating job

### PR DESCRIPTION
**Pull request by weezer**
adding leapfrog-multinode-aio job into nightly gating. working in progress.

Issue: [AR-106](https://rpc-openstack.atlassian.net/browse/AR-106)

### CODE REVIEW

File: rpc_jobs/leapfrog_multi_node_aio.yml [hunk 1]
```diff
@@ -0,0 +1,136 @@
+- project:
```
Comment (mkam) [inline rpc_jobs/leapfrog_multi_node_aio.yml]: This might be better as an addition to the exiting multi_node_aio.yml file instead of a separate one so that it's closer to what we do for AIO leapfrogs.

File: rpc_jobs/leapfrog_multi_node_aio.yml [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+                "DEPLOY_AIO=no",
+                "DEPLOY_MAAS=no"
+                ]
+            ) // deploy_sh
+            multi_node_aio_prepare.multi_node_aio_networking()
+            dir("rpc-gating"){{
+              // Checkout rpc-gating repo to avoid race conditions in parallel steps
+              git branch: env.RPC_GATING_BRANCH, url: env.RPC_GATING_REPO
+            }}
+            deploy.upgrade_leapfrog(vm: "deploy1", environment_vars: environment_vars)
```
Comment (mkam) [inline rpc_jobs/leapfrog_multi_node_aio.yml]: You'll want to run the holland/maas/kibana/tempest tests before the upgrade as well as after it.

Comment (weezer) [inline rpc_jobs/leapfrog_multi_node_aio.yml]: does the kibana and mass work the same for both kilo and newtown?

File: rpc_jobs/multi_node_aio.yml [hunk 3]
```diff
@@ -34,6 +34,24 @@
       - 'OnMetal-Multi-Node-AIO_{series}-{context}-{trigger}'
       - 'OnMetal-Multi-Node-AIO-Merge-Trigger_{series}'
 
+- project:
+    name: 'Leapfrog-Multi-Node-AIO-Jobs'
+    series:
+      - kilo:
+          branch: kilo
+          OPENSTACK_ANSIBLE_BRANCH: "stable/kilo"
+          TEST_SET: ""
```
Comment (mkam) [inline rpc_jobs/multi_node_aio.yml]: The job currently isn't building because USER_VARS isn't set.  Also, we don't use TEST_SET anymore, so you'll want to change TEST_SET to something like:

```
USER_VARS: |
    tempest_test_sets: 'all'
```
We run `all` for newton, though I'm not sure what set will pass against kilo.
 
https://github.com/weezer/rpc-gating/blob/5cf7600c649a40e22a351e9c5e7b1060ff7acb03/rpc_jobs/multi_node_aio.yml#L10-L11

Comment (weezer) [inline rpc_jobs/multi_node_aio.yml]: ok, i will try to wrap it up with hugh's comments

File: .gitignore [hunk 4]
```diff
@@ -1,7 +1,8 @@
+.idea/
```
Comment (git-harry) [inline .gitignore]: What is .idea for?

File: pipeline_steps/deploy.groovy [hunk 5]
```diff
@@ -55,11 +55,22 @@ def upgrade(String stage_name, String upgrade_script, List env_vars) {
           sh "git reset --hard"
         }
         common.prepareRpcGit()
```
Comment (git-harry) [inline pipeline_steps/deploy.groovy]: If `vm` is set, how does the repo get updated on it? On the initial deployment the repo appears to get copied across, https://github.com/rcbops/rpc-gating/blob/master/pipeline_steps/multi_node_aio_prepare.groovy#L53, but I can't see anything similar here with the upgrade step.

File: pipeline_steps/deploy.groovy [hunk 6]
```diff
@@ -55,11 +55,22 @@ def upgrade(String stage_name, String upgrade_script, List env_vars) {
...
+            """
+          } // dir
+        } else {
+            export_vars = ""
+            for ( e in environment_vars ) {
+              export_vars += "export ${e}; "
+            }
+            sh """#!/bin/bash
+            sudo ssh -T -oStrictHostKeyChecking=no ${vm} \
+              '${export_vars} cd /opt/rpc-openstack; scripts/leapfrog/ubuntu14-leapfrog.sh'
```
Comment (git-harry) [inline pipeline_steps/deploy.groovy]: I think this is supposed to be:

```
'${export_vars} cd /opt/rpc-openstack; scripts/$upgrade_script'
```

File: rpc_jobs/multi_node_aio.yml [hunk 7]
```diff
@@ -34,6 +34,24 @@
       - 'OnMetal-Multi-Node-AIO_{series}-{context}-{trigger}'
       - 'OnMetal-Multi-Node-AIO-Merge-Trigger_{series}'
 
+- project:
+    name: 'Leapfrog-Multi-Node-AIO-Jobs'
+    series:
+      - kilo:
```
Comment (git-harry) [inline rpc_jobs/multi_node_aio.yml]: I think that is worse, if all the jobs are the other way round lets be consistent.

File: rpc_jobs/multi_node_aio.yml [hunk 8]
```diff
@@ -190,3 +208,154 @@
...
+    properties:
+      - build-discarder:
+          days-to-keep: 7
+      - rpc-openstack-github
+    concurrent: true
+    triggers:
+      - timed: "{CRON}"
+    parameters:
+      - kibana_selenium_params:
+        KIBANA_SELENIUM_BRANCH: "{branch}"
```
Comment (git-harry) [inline rpc_jobs/multi_node_aio.yml]: Missing indentation

File: rpc_jobs/multi_node_aio.yml [hunk 9]
```diff
@@ -190,3 +208,154 @@
...
+      - rpc_gating_params
+      - instance_params:
+          IMAGE: "OnMetal - Ubuntu 14.04 LTS (Trusty Tahr)"
+          FLAVOR: "onmetal-io1"
+          REGION: "IAD"
+      - rpc_repo_params:
+          RPC_BRANCH: "{branch}"
+      - osa_ops_params:
+          DEFAULT_IMAGE: "3.13.0-98"
```
Comment (git-harry) [inline rpc_jobs/multi_node_aio.yml]: That looks more like a kernel than an image.

File: rpc_jobs/multi_node_aio.yml [hunk 10]
```diff
@@ -190,3 +208,154 @@
...
+              //}}, kilo has its own horizon integration project from openstack
+             // "kibana": {{
+             //   kibana.kibana(env.KIBANA_SELENIUM_BRANCH, "deploy1")
+             // }}, dont think the kibana has kilo branch.
+              "holland": {{
+                holland.holland("deploy1")
+              }}
+            ) // parallel
+            deploy.upgrade_leapfrog(vm: "deploy1", environment_vars: environment_vars)
+            parallel(
```
Comment (git-harry) [inline rpc_jobs/multi_node_aio.yml]: This job seems to be introducing further duplication, for example this test section. I would be nice to limit that where reasonable.

File: .gitignore (hunk 4 above)
Comment (weezer) [inline .gitignore]: pycharm's configuration file, will be automatically git with, so have to ignore it.

File: rpc_jobs/multi_node_aio.yml (hunk 10 above)
Comment (weezer) [inline rpc_jobs/multi_node_aio.yml]: we want to test it after kilo been deployed, and test it again after leaping, tricky thing is, kibana doesn't have a kilo branch, and horizon has totally different test methods for kilo and newton.

File: pipeline_steps/deploy.groovy (hunk 5 above)
Comment (weezer) [inline pipeline_steps/deploy.groovy]: the multi-node-aio repo been copied into a virtual machine names deploy1, not the rpc-openstack, the baremetal will be just like a switch/router, and all those jobs are done within the virtual machines.

File: rpc_jobs/multi_node_aio.yml (hunk 10 above)
Comment (odyssey4me) [inline rpc_jobs/multi_node_aio.yml]: Don't bother with tests you do not already have. All you need to test for Kilo is the basic functionality. The work to implement some VM's, volumes, etc and monitor them through the leap upgrade process is enough of a test to validate that the installation is functional.

File: pipeline_steps/deploy.groovy (hunk 5 above)
Comment (hughsaunders) [inline pipeline_steps/deploy.groovy]: @weezer I think the problem @git-harry is pointing out is that common.prepareRpcGit will prepare /opt/rpc-openstack on the host. If that is not copied into deploy1 vm, then there is no point in running the upgrade script on deploy1 as it will still have the same version of rpc-openstack that was deployed the first time.

Comment (hughsaunders) [inline pipeline_steps/deploy.groovy]: @weezer I think the problem @git-harry is pointing out is that common.prepareRpcGit will prepare /opt/rpc-openstack on the host. If that is not copied into deploy1 vm, then there is no point in running the upgrade script on deploy1 as it will still have the same version of rpc-openstack that was deployed the first time.

**Comment by evrardjp** [comment]
I will review when I have finished my backlog and all the other comments have been addressed.
Sorry for the inconvenience.

File: rpc_jobs/multi_node_aio.yml (hunk 7 above)
Comment (evrardjp) [inline rpc_jobs/multi_node_aio.yml]: Agreed with @git-harry . Series should be newton14.1, even if it's a new project.

File: pipeline_steps/multi_node_aio_prepare.groovy [hunk 11]
```diff
@@ -57,12 +57,10 @@ def prepare() {
       set -xe
       sudo cp /etc/openstack_deploy/user_variables.yml /etc/openstack_deploy/user_variables.yml.bak
       sudo cp -R /opt/rpc-openstack/openstack-ansible/etc/openstack_deploy /etc
+      sudo cp /opt/rpc-openstack/rpcd/etc/openstack_deploy/user_*.yml /etc/openstack_deploy
```
Comment (evrardjp) [inline pipeline_steps/multi_node_aio_prepare.groovy]: Why this change? Could you please document it?

File: rpc_jobs/multi_node_aio.yml [hunk 12]
```diff
@@ -190,3 +206,145 @@
...
+            Pipeline stages to run CSV. Note that this list does not influence execution order.
+            Options:
+              Allocate Resources
+              Connect Slave
+              Prepare Multi-Node AIO
+              Prepare RPC Configs
+              Deploy RPC w/ Script
+              Prepare MaaS
+              Setup MaaS
+              Verify MaaS
```
Comment (evrardjp) [inline rpc_jobs/multi_node_aio.yml]: Not sure we need to verify maas, tempest tests and holland  before leapfrog. It should be already tested elsewhere. (Kilo deploy on multi node). If kilo deploy on multi node doesn't exist yet, we should build it.

File: rpc_jobs/multi_node_aio.yml (hunk 10 above)
Comment (evrardjp) [inline rpc_jobs/multi_node_aio.yml]: agreed with @odyssey4me

**Comment by evrardjp** [comment]
Can't we extend the existing multi node jobs to add leapfrog, like we did for AIO?

**Comment by weezer** [comment]
testing it right now, will you ping you guys and paste the test result here first.

File: rpc_jobs/multi_node_aio.yml [hunk 13]
```diff
@@ -190,3 +208,129 @@
...
+            prepareRpcGit(env.UPGRADE_TO_REF)
+            sh """/bin/bash
+            scp -r -o StrictHostKeyChecking=no /opt/rpc-openstack deploy1:/opt/
+            EOF
+            """
+            deploy.upgrade_leapfrog(vm: "deploy1", environment_vars: environment_vars)
+            dir("rpc-gating"){{
+              // Checkout rpc-gating repo to avoid race conditions in parallel steps
+              git branch: env.RPC_GATING_BRANCH, url: env.RPC_GATING_REPO
+              """
```
Comment (mkam) [inline rpc_jobs/multi_node_aio.yml]: This needs to be removed.

File: pipeline_steps/deploy.groovy [hunk 14]
```diff
@@ -36,7 +36,7 @@ def deploy_sh(Map args) {
             }
             sh """#!/bin/bash
             sudo ssh -T -oStrictHostKeyChecking=no ${args.vm} \
-              '${export_vars} cd /opt/rpc-openstack; scripts/deploy.sh'
+              '${export_vars} cd /opt/rpc-openstack; FORKS=5; scripts/deploy.sh'
```
Comment (odyssey4me) [inline pipeline_steps/deploy.groovy]: Why are the FORKS being fixed to this value?

Comment (weezer) [inline pipeline_steps/deploy.groovy]: hi pls ignoring it, because the vm only has 4 cores and i want to change the ansible fork number to 5 for stability. i should closed this ticket first then reopen it

**Comment by weezer** [comment]
abondon
